apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+        private static void ValidateCustomMetadataKey(string key)
+        {
+            if (!KeyRegex.IsMatch(key))
+            {
+                throw new ArgumentException("Invalid custom metadata key.", "key");
+            }
+        }
+
+        /// <inheritdoc/>
+        public override TokenResponse GetAccessToken() => GetResult(GetAccessTokenAsync);
+
+        /// <inheritdoc/>
+        public override async Task<TokenResponse> GetAccessTokenAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // TODO: Most of this is taken from Google.Apis.Auth.OAuth2.ComputeCredential.RequestAccessTokenAsync. Perhaps move that to a helper which can be shared.
+            var response = await RequestMetadata("instance/service-accounts/default/token", cancellationToken).ConfigureAwait(false);
+
+            // Read the response.
+            var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            if (!response.IsSuccessStatusCode)
+            {
+                try
+                {
+                    var error = NewtonsoftJsonSerializer.Instance.Deserialize<TokenErrorResponse>(content);
+                    throw new TokenResponseException(error);
+                }
+                // Json.NET deserialize object method doesn't include documentation about which exceptions can
+                // be thrown, that's why we just catch a general exception here.
+                catch (Exception ex)
+                {
+                    Logger.Error(ex, "Exception was caught when deserializing TokenErrorResponse. Content is: {0}",
+                        content);
+                    var error = "Server response does not contain a JSON object. Status code is: "
+                        + response.StatusCode;
+                    throw new TokenResponseException(new TokenErrorResponse
+                    {
+                        Error = error
+                    });
+                }
+            }
+
+            // Gets the token and sets its issued time.
+            var newToken = NewtonsoftJsonSerializer.Instance.Deserialize<TokenResponse>(content);
+            newToken.Issued = DateTime.UtcNow; // TODO: This was using an IClock before.
+            return newToken;
+        }
+
+        /// <inheritdoc/>
+        public override string GetCustomInstanceMetadata(string key) => GetResult(token => GetCustomInstanceMetadataAsync(key, token));
+
+        /// <inheritdoc/>
+        public override async Task<string> GetCustomInstanceMetadataAsync(string key, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            ValidateCustomMetadataKey(key);
+            var response = await RequestMetadata("instance/attributes/" + key, cancellationToken).ConfigureAwait(false);
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
```

Comment:
I think I do in this case since I'm awaiting earlier in the method, so it must be marked async, which means I need to return a string.